File: csharp/Scorers.cs
// EvalForge scorers: turn (output, expected) into a 0..1 score.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace EvalForge;

public static class Scorers
{
    public static double ExactMatch(string output, string expected)
        => output.Trim() == expected.Trim() ? 1.0 : 0.0;

    public static double Contains(string output, string expected)
        => output.ToLowerInvariant().Contains(expected.Trim().ToLowerInvariant()) ? 1.0 : 0.0;

    public static double RegexMatch(string output, string pattern)
        => Regex.IsMatch(output, pattern) ? 1.0 : 0.0;

    /// <summary>
    /// Lightweight, dependency-free proxy for semantic closeness: token overlap +
    /// sequence ratio (Ratcliff/Obershelp, matching Python's difflib). Swap for
    /// embedding cosine or an LLM-as-judge in production; the interface is identical.
    /// </summary>
    public static double SemanticSimilarity(string output, string expected)
    {
        var a = output.ToLowerInvariant().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        var b = expected.ToLowerInvariant().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        if (a.Length == 0 || b.Length == 0) return 0.0;
        var sa = new HashSet<string>(a);
        var sb = new HashSet<string>(b);
        var inter = sa.Count(x => sb.Contains(x));
        var union = new HashSet<string>(sa);
        union.UnionWith(sb);
        var overlap = (double)inter / union.Count;
        var ratio = SequenceRatio(output.ToLowerInvariant(), expected.ToLowerInvariant());
        return Math.Round(0.5 * overlap + 0.5 * ratio, 4);
    }

    // --- difflib.SequenceMatcher.ratio() (Ratcliff/Obershelp), no autojunk ---

    private static double SequenceRatio(string a, string b)
    {
        if (a.Length + b.Length == 0) return 1.0;
        var b2j = new Dictionary<char, List<int>>();
        for (var j = 0; j < b.Length; j++)
        {
            if (!b2j.TryGetValue(b[j], out var list)) { list = new List<int>(); b2j[b[j]] = list; }
            list.Add(j);
        }
        var matches = TotalMatches(a, b, b2j, 0, a.Length, 0, b.Length);
        return 2.0 * matches / (a.Length + b.Length);
    }

    private static int TotalMatches(string a, string b, Dictionary<char, List<int>> b2j,
        int alo, int ahi, int blo, int bhi)
    {
        var (i, j, k) = FindLongestMatch(a, b2j, alo, ahi, blo, bhi);
        if (k == 0) return 0;
        return k
            + TotalMatches(a, b, b2j, alo, i, blo, j)
            + TotalMatches(a, b, b2j, i + k, ahi, j + k, bhi);
    }

    private static (int, int, int) FindLongestMatch(string a, Dictionary<char, List<int>> b2j,
        int alo, int ahi, int blo, int bhi)
    {
        int besti = alo, bestj = blo, bestsize = 0;
        var j2len = new Dictionary<int, int>();
        for (var i = alo; i < ahi; i++)
        {
            var newj2len = new Dictionary<int, int>();
            if (b2j.TryGetValue(a[i], out var js))
            {
                foreach (var j in js)
                {
                    if (j < blo) continue;
                    if (j >= bhi) break;
                    var k = (j2len.TryGetValue(j - 1, out var prev) ? prev : 0) + 1;
                    newj2len[j] = k;
                    if (k > bestsize)
                    {
                        besti = i - k + 1;
                        bestj = j - k + 1;
                        bestsize = k;
                    }
                }
            }
            j2len = newj2len;
        }
        return (besti, bestj, bestsize);
    }

    private static readonly HashSet<string> Names = new() { "exact_match", "contains", "regex", "semantic" };

    public static bool IsKnown(string name) => Names.Contains(name);

    /// <summary>Score by scorer name. <paramref name="arg"/> is the pattern for
    /// "regex", otherwise the expected string.</summary>
    public static double Score(string name, string output, string arg) => name switch
    {
        "exact_match" => ExactMatch(output, arg),
        "contains" => Contains(output, arg),
        "regex" => RegexMatch(output, arg),
        "semantic" => SemanticSimilarity(output, arg),
        _ => throw new ArgumentException(
            $"unknown scorer '{name}'. available: [{string.Join(", ", Names)}]"),
    };
}

File: rust/src/scorers.rs
//! EvalForge scorers: turn a target output and an argument into a 0..1 score.

use regex::Regex;
use std::collections::HashMap;
use std::collections::HashSet;

/// A scorer turns a target output and a single argument (the expected string,
/// or the pattern for the regex scorer) into a score in the range 0..1.
pub type Scorer = fn(&str, &str) -> f64;

/// Scores 1.0 when output and expected are equal after trimming surrounding
/// whitespace, else 0.0.
pub fn exact_match(output: &str, expected: &str) -> f64 {
    if output.trim() == expected.trim() {
        1.0
    } else {
        0.0
    }
}

/// Scores 1.0 when expected (trimmed and lower-cased) appears anywhere in
/// output (lower-cased), else 0.0.
pub fn contains(output: &str, expected: &str) -> f64 {
    let needle = expected.trim().to_lowercase();
    if output.to_lowercase().contains(needle.as_str()) {
        1.0
    } else {
        0.0
    }
}

/// Scores 1.0 when pattern matches anywhere in output, else 0.0. The pattern
/// supports inline flags such as `(?i)`; an invalid pattern panics, mirroring
/// the Python reference raising `re.error`.
pub fn regex_match(output: &str, pattern: &str) -> f64 {
    let re = Regex::new(pattern).expect("invalid regex pattern");
    if re.is_match(output) {
        1.0
    } else {
        0.0
    }
}

/// A lightweight, dependency-free proxy for semantic closeness: the mean of
/// Jaccard token overlap and difflib's sequence ratio, rounded to four
/// decimals. Not a real embedding model; the scorer interface is identical so
/// it can be swapped for one in production.
pub fn semantic_similarity(output: &str, expected: &str) -> f64 {
    let lo = output.to_lowercase();
    let le = expected.to_lowercase();
    let a: Vec<&str> = lo.split_whitespace().collect();
    let b: Vec<&str> = le.split_whitespace().collect();
    if a.is_empty() || b.is_empty() {
        return 0.0;
    }
    let sa: HashSet<&str> = a.iter().copied().collect();
    let sb: HashSet<&str> = b.iter().copied().collect();
    let inter = sa.intersection(&sb).count();
    let union = sa.union(&sb).count();
    let overlap = inter as f64 / union as f64;
    let ratio = sequence_ratio(&lo, &le);
    round4(0.5 * overlap + 0.5 * ratio)
}

/// Rounds v to four decimal places using round-half-to-even, reproducing the
/// value Python's `round(v, 4)` yields for the same input.
pub fn round4(v: f64) -> f64 {
    format!("{v:.4}").parse::<f64>().unwrap()
}

/// Reproduces `difflib.SequenceMatcher(None, a, b).ratio()`: it is `2*M/T` where
/// `T` is the combined length of both strings and `M` is the total size of the
/// matching blocks found by the Ratcliff/Obershelp algorithm. Comparison is per
/// Unicode code point, and the autojunk heuristic is not applied (it only
/// affects sequences of 200+ elements, well beyond the short strings scored
/// here).
fn sequence_ratio(a: &str, b: &str) -> f64 {
    let ra: Vec<char> = a.chars().collect();
    let rb: Vec<char> = b.chars().collect();
    if ra.is_empty() && rb.is_empty() {
        return 1.0;
    }
    let mut b2j: HashMap<char, Vec<usize>> = HashMap::new();
    for (j, &ch) in rb.iter().enumerate() {
        b2j.entry(ch).or_default().push(j);
    }
    let matches = total_matches(&ra, &b2j, 0, ra.len(), 0, rb.len());
    2.0 * matches as f64 / (ra.len() + rb.len()) as f64
}

fn total_matches(
    a: &[char],
    b2j: &HashMap<char, Vec<usize>>,
    alo: usize,
    ahi: usize,
    blo: usize,
    bhi: usize,
) -> usize {
    let (i, j, k) = find_longest_match(a, b2j, alo, ahi, blo, bhi);
    if k == 0 {
        return 0;
    }
    k + total_matches(a, b2j, alo, i, blo, j) + total_matches(a, b2j, i + k, ahi, j + k, bhi)
}

fn find_longest_match(
    a: &[char],
    b2j: &HashMap<char, Vec<usize>>,
    alo: usize,
    ahi: usize,
    blo: usize,
    bhi: usize,
) -> (usize, usize, usize) {
    let mut besti = alo;
    let mut bestj = blo;
    let mut bestsize = 0usize;
    let mut j2len: HashMap<usize, usize> = HashMap::new();
    for (i, item) in a.iter().enumerate().take(ahi).skip(alo) {
        let mut newj2len: HashMap<usize, usize> = HashMap::new();
        if let Some(js) = b2j.get(item) {
            for &j in js {
                if j < blo {
                    continue;
                }
                if j >= bhi {
                    break;
                }
                let prev = if j == 0 {
                    0
                } else {
                    *j2len.get(&(j - 1)).unwrap_or(&0)
                };
                let k = prev + 1;
                newj2len.insert(j, k);
                if k > bestsize {
                    besti = i + 1 - k;
                    bestj = j + 1 - k;
                    bestsize = k;
                }
            }
        }
        j2len = newj2len;
    }
    (besti, bestj, bestsize)
}

/// The built-in scorer names, in registration order.
pub const SCORER_NAMES: [&str; 4] = ["exact_match", "contains", "regex", "semantic"];

/// Returns the names of the built-in scorers.
pub fn scorer_names() -> Vec<String> {
    SCORER_NAMES.iter().map(|s| s.to_string()).collect()
}

/// Reports whether name is a built-in scorer.
pub fn is_known_scorer(name: &str) -> bool {
    SCORER_NAMES.contains(&name)
}

/// Returns the scorer registered under name, or an error if no such scorer
/// exists, mirroring the Python reference's `get_scorer`.
pub fn get_scorer(name: &str) -> Result<Scorer, String> {
    match name {
        "exact_match" => Ok(exact_match),
        "contains" => Ok(contains),
        "regex" => Ok(regex_match),
        "semantic" => Ok(semantic_similarity),
        _ => Err(format!(
            "unknown scorer '{name}'. available: {SCORER_NAMES:?}"
        )),
    }
}
